staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -148,7 +153,19 @@ func (t *envelopeTransformer) TransformFromStorage(ctx context.Context, data []b
 		}
 	}
 
-	return transformer.TransformFromStorage(ctx, encryptedObject.EncryptedData, dataCtx)
+	out, stale, err := transformer.TransformFromStorage(ctx, encryptedObject.EncryptedData, dataCtx)
+
+	// Check keyID freshness in addition to data staleness
+	if !stale {
+		keyID, err := t.keyIDGetter(ctx)
+		if err != nil {
+			return nil, false, err
+		} else {
+			stale = encryptedObject.KeyID != keyID
+		}
+	}
+
+	return out, stale, err
```

Comment:
(once you return early we know there is no error)

```suggestion
	return out, stale, nil
```